# **UTS**

Lakukan klasifikasi text abstrak dengan naive bayes / KNN menggunakan representasi text TF-IDF dengan reduksi dimensi PCA!







In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## **1. Proses Load Data**

In [2]:
import pandas as pd
import numpy as np

In [3]:
# Create dataset from Google Drive
dataset_url = "https://raw.githubusercontent.com/AlifnurFathurrahmanPrasodjo/dataFolder/main/dataMining/dataset-baru.csv"
df = pd.read_csv(dataset_url)
df

Unnamed: 0                                            Abstrak   Kategori
0             0  Sistem  informasi  akademik  (SIAKAD) merupaka...        RPL
1             1  Berjalannya koneksi jaringan komputer dengan l...        RPL
2             2  Web server adalah sebuah perangkat lunak serve...        RPL
3             3  Penjadwalan  kuliah  di  Perguruan  Tinggi  me...  KOMPUTASI
4             4  Seiring perkembangan teknologi yang ada diduni...        RPL
..          ...                                                ...        ...
800         804  Investasi saham selama ini memiliki resiko ker...  KOMPUTASI
801         805  Information Retrieval (IR) merupakan pengambil...  KOMPUTASI
802         806  Klasifikasi citra merupakan proses pengelompok...  KOMPUTASI
803         807  Identifikasi atribut pejalan kaki merupakan sa...  KOMPUTASI
804         808  Topik deteksi objek telah menarik perhatian ya...  KOMPUTASI

[805 rows x 3 columns]

In [4]:
text = df['Abstrak']
X = text
Y = df.values[:, -1]

## **2. Proses NPL Data**

In [5]:
!pip install sastrawi
!pip install swifter

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 209.7/209.7 kB 17.0 MB/s eta 0:00:00
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 830.9/830.9 kB 34.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 54.9 MB/s eta 0:00:00
  Created wheel for swifter: filename=swifter-1.3.4-py3-none-any.whl size=16321 sha256=d90d1c08179ddaa1bbde6a9c393721074a7532ed8191aaab1590f377b6b29988
  Stored in directory: /root/.cache/pip/wheels/2b/5e/f2/3931524f702ffd03309e96d35ee2fbf9c61c27377511ee8d4c
Successfully built swifter


In [6]:
import pandas as pd
import re
import string
import numpy as np
import nltk
import swifter
from nltk.corpus import stopwords
from Sastrawi.Stemmer.StemmerFactory import StemmerFactory
import matplotlib.pyplot as plt

In [7]:
def case_folding(comment):
    comment = comment.lower()
    return comment
def linkNormalize(text):
    text = re.sub(r"\s—\s", "", text)
    text = re.sub(r"http\S+", "", text)
    return text
symbols = ",!\"#$%&()*+-.…/:;<=>?@[\]^_`{|}~\n"
def cleaning_simbol(comment):
    for i in symbols:
        comment = np.char.replace(comment, i, ' ')
    return comment
def tokenize(comment):
    return comment.split()

nltk.download('stopwords')
list_stopwords = stopwords.words('indonesian')
def stopword_removal(comment):
    return [word for word in comment if word not in list_stopwords]

factory = StemmerFactory()
stemmer = factory.create_stemmer()
def stemmingNormalize(comment):
#     print(comment)
    kata = []
    for term in comment:
        term = stemmer.stem(term)
        kata.append(term)
    sentence = ""
    for i in kata:
      sentence += i
      sentence += " "
    return sentence

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


### 1. Case - Folding Data
> Proses case folding bertujuan untuk membuat seluruh kata menjadi huruf kecil semua.



In [8]:
X = X.apply(case_folding)
k=0
for i in X:
    #print(k,i)
    k += 1

### 2. Cleaning - Text Data


> 1. linkNormalize, bertujuan untuk menghilangkan alamat link pada data
> 2. cleaning_simbol, bertujuan untuk menghilangkan simbol yang tidak perlu pada data





In [9]:
X = X.apply(linkNormalize)
k=0
for i in X:
    #print(k,i)
    k += 1


In [10]:
X = X.apply(cleaning_simbol)
k=0
for i in X:
    text = i.split()
    #print(k,i)
    k += 1

### 3. Tokenize
> Proses tokenizing bertujuan untuk memecah kalimat menjadi kata.



In [11]:
X = X.apply(tokenize)
X

0      [sistem, informasi, akademik, siakad, merupaka...
1      [berjalannya, koneksi, jaringan, komputer, den...
2      [web, server, adalah, sebuah, perangkat, lunak...
3      [penjadwalan, kuliah, di, perguruan, tinggi, m...
4      [seiring, perkembangan, teknologi, yang, ada, ...
                             ...                        
800    [investasi, saham, selama, ini, memiliki, resi...
801    [information, retrieval, ir, merupakan, pengam...
802    [klasifikasi, citra, merupakan, proses, pengel...
803    [identifikasi, atribut, pejalan, kaki, merupak...
804    [topik, deteksi, objek, telah, menarik, perhat...
Name: Abstrak, Length: 805, dtype: object

### 4. Stopword Removal
> Proses stopword removal bertujuan untuk menghilangkan kata yang tidak perlu.



In [12]:
X = X.apply(stopword_removal)
k=0
for i in X:
    #print(k,i)
    k += 1

### 5. Stemming
> Proses stemming bertujuan untuk membuat kata yang berimbuhan menjadi kata dasar.



In [13]:
X = X.swifter.apply(stemmingNormalize)
X

Pandas Apply:   0%|          | 0/805 [00:00<?, ?it/s]

0      sistem informasi akademik siakad sistem inform...
1      jalan koneksi jaring komputer lancar ganggu ha...
2      web server perangkat lunak server fungsi terim...
3      jadwal kuliah guru kompleks masalah variabel t...
4      iring kembang teknologi dunia muncul teknologi...
                             ...                        
800    investasi saham milik resiko rugi gera harga s...
801    information retrieval ir ambil informasi simpa...
802    klasifikasi citra proses kelompok piksel citra...
803    identifikasi atribut pejal kaki salah teliti k...
804    topik deteksi objek tarik perhati kembang tekn...
Name: Abstrak, Length: 805, dtype: object

## **3. Proses Representasi Data Menggunakan Algoritma TF-IDF**


> Term Frequency — Inverse Document Frequency atau TF — IDF adalah suatu metode algoritma yang berguna untuk menghitung bobot setiap kata yang umum digunakan. Metode ini akan menghitung nilai Term Frequency (TF) dan Inverse Document Frequency (IDF) pada setiap token (kata) di setiap dokumen. Secara sederhana, metode TF-IDF digunakan untuk mengetahui berapa sering suatu kata muncul di dalam dokumen.



In [14]:
from sklearn.feature_extraction.text import TfidfVectorizer
# )
tfidf_vectorizer = TfidfVectorizer()
tfidf_separate = tfidf_vectorizer.fit_transform(X)

df_tfidf = pd.DataFrame(
    tfidf_separate.toarray(), columns=tfidf_vectorizer.get_feature_names_out(), index=df.index
)
X = df_tfidf.values
df_tfidf

00  000  00001  0001  000505341  000506436  000792353  00086  00089  \
0    0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
1    0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
2    0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
3    0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
4    0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
..   ...  ...    ...   ...        ...        ...        ...    ...    ...   
800  0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
801  0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
802  0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
803  0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
804  0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   

     001  ...  zara  zat  zcz   zf  zona  zone  zoning  zoom  zucara  zungu  
0    0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  
1    0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  
2    0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  
3    0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  
4    0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  
..   ...  ...   ...  ...  ...  ...   ...   ...     ...   ...     ...    ...  
800  0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  
801  0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  
802  0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  
803  0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  
804  0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  

[805 rows x 6570 columns]

## **4. Proses Splitting Data**

> Dalam proses splitting data ini dibagi menjadi 2 bagian, yakni data training dan data testing. Data training menggunakan sebanyak 80% dari keseluruhan data sedangkan data testing menggunakan sebanyak 20% dari keseluruhan data



In [15]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split( 
    X, Y, test_size = 0.2, random_state = 100)

y_train

array(['KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI',
       'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI',
       'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'RPL', 'KOMPUTASI',
       'RPL', 'KOMPUTASI', 'RPL', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'RPL',
       'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'RPL',
       'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI',
       'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'RPL',
       'KOMPUTASI', 'KOMPUTASI', 'RPL', 'RPL', 'KOMPUTASI', 'KOMPUTASI',
       'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI',
       'RPL', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI',
       'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI',
       'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI',
       'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI',
       'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI',
       'KOMPUTA

## **5. Proses Reduksi Data Menggunakan Metode PCA dan Proses Klasifikasi Data menggunakan metode KNN (K-Nearest Neighbor)**

> Principal Component Analysis (PCA) adalah salah satu metode reduksi dimensi pada machine learning. PCA akan memilih œvariabel-variabel yang mampu menjelaskan sebagian besar variabilitas data. PCA mengurangi dimensi dengan membentuk variabel-variabel baru

> KNN merupakan algoritma klasifikasi yang paling sederhana dalam mengklasifikasikan sebuah gambar kedalam sebuah label. Metode ini mudah dipahami dibandingkan metode lain karena mengklasifikasikan berdasarkan jarak terdekat dengan objek lain (tetangga).





### Tahap Uji Coba

1. Uji coba pengecekan hasil akurasi data menggunakan metode klasifikasi KNN (K-Nearest Neighbor) terhadap data **sebelum** dilakukannya proses reduksi data menggunakan metode PCA.

In [16]:
from sklearn.neighbors import KNeighborsClassifier
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(X_train, y_train)

KNeighborsClassifier()

In [17]:
from sklearn.metrics import accuracy_score
y_pred = classifier.predict(X_test)
print (" KNN Accuracy : ",
    accuracy_score(y_test,y_pred)*100)

 KNN Accuracy :  83.85093167701864



> **Sebelum** dilakukannya proses reduksi data menggunakan metode PCA, pengecekan hasil akurasi data menggunakan metode klasifikasi KNN dengan jumlah fitur data sebanyak **6570** menghasilkan akurasi sebesar **83.85**.



2. Uji coba pengecekan hasil akurasi data menggunakan metode klasifikasi KNN (K-Nearest Neighbor) terhadap data **setelah** dilakukannya proses reduksi data menggunakan metode PCA.
> 



In [18]:
from sklearn.decomposition import PCA

pca = PCA(n_components =10)

X_train = pca.fit_transform(X_train)
X_test = pca.transform(X_test)

pca.explained_variance_ratio_

array([0.02192352, 0.01584224, 0.01479794, 0.01248872, 0.01058649,
       0.00931853, 0.00844545, 0.00801684, 0.00780416, 0.00727916])

In [19]:
from sklearn.neighbors import KNeighborsClassifier
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(X_train, y_train)

KNeighborsClassifier()

In [20]:
from sklearn.metrics import accuracy_score
y_pred = classifier.predict(X_test)
print (" KNN Accuracy : ",
    accuracy_score(y_test,y_pred)*100)

 KNN Accuracy :  79.5031055900621



> **Setelah** dilakukannya proses reduksi data menggunakan metode PCA dengan jumlah n_components sebanyak 10, pengecekan hasil akurasi data menggunakan metode klasifikasi KNN dengan jumlah fitur data sebanyak **10** menghasilkan akurasi sebesar **82.60**.



## **Kesimpulan**

> Dari tahap uji coba yang dilakukan pada proses reduksi data menggunakan metode PCA dan proses klasifikasi data menggunakan metode KNN (K-Nearest Neighbor), dapat disimpulkan bahwa dengan diterapkannya metode PCA dapat mempercepat proses komputasi klasifikasi karena data yang sebelumnya memiliki fitur 6570 menjadi 10 fitur. Meskipun hasil akurasi turun sebanyak **83.85 - 82.60 = 1.25**, namun dapat mempercepat proses komputasi klasifikasi.



## **6. Proses Reduksi Data Menggunakan Metode PCA dan Proses Klasifikasi Data menggunakan metode K-Means (Jumlah Cluster 2)**

> K-means clustering dapat didefinisikan sebagai algoritma iteratif yang membagi kumpulan data (dataset) yang tidak berlabel menjadi k kluster yang berbeda sedemikian rupa sehingga setiap kumpulan data hanya dimiliki oleh satu kelompok yang memiliki properti serupa.




In [21]:
from sklearn.cluster import KMeans
kmeans = KMeans(n_clusters=2, random_state=42)
kmeans.fit(X)

KMeans(n_clusters=2, random_state=42)

In [22]:
center = kmeans.cluster_centers_
center

array([[ 6.65956073e-04,  6.56155462e-04, -1.35525272e-20, ...,
         0.00000000e+00, -5.42101086e-20, -5.42101086e-20],
       [ 8.27791219e-04,  1.27823885e-03,  1.17526274e-04, ...,
         4.03049443e-04,  1.40407746e-04,  1.40407746e-04]])

In [23]:
y_pred = kmeans.labels_


data = kmeans.labels_
res = []
for i in data:
    if i == 0:
        res.append("RPL")
    else:
        res.append("KOMPUTASI")
print(res)

['KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMP

In [24]:
y_true = df['Kategori'].values.tolist()
print(df['Kategori'].values.tolist())

['RPL', 'RPL', 'RPL', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'RPL', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'RPL', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'KOMPUTASI', 'K

In [25]:
n = len(y_true)
for i in range(n):
    print(res[i],"\t : ",y_true[i])

KOMPUTASI 	 :  RPL
KOMPUTASI 	 :  RPL
KOMPUTASI 	 :  RPL
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  RPL
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
RPL 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  RPL
KOMPUTASI 	 :  RPL
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
RPL 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
RPL 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
RPL 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
RPL 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  RPL
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  RPL
KOMPUTASI 	 :  KOMPUTASI
RPL 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
RPL 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  RPL
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUTASI
KOMPUTASI 	 :  KOMPUT

In [26]:
from sklearn.metrics import accuracy_score
print ("Accuracy : ",
    accuracy_score(y_true, res)*100)

Accuracy :  53.16770186335403
